diff --git a/rt-test/alloc-test-bpa.cc b/rt-test/alloc-test-bpa.cc
index 237273c..10b00ab 100644
--- a/rt-test/alloc-test-bpa.cc
+++ b/rt-test/alloc-test-bpa.cc
@@ -7,6 +7,7 @@
 #include "user/bmalloc.h"
 
 void main() {
+    return;
   for (int i = 4096; i < 1 << 16; ++i) {
     auto block = block_alloc(i, 4096);
     assert(block.begin);
diff --git a/rt-test/alloc-test-mpa.cc b/rt-test/alloc-test-mpa.cc
index 15d9756..a6eb8c6 100644
--- a/rt-test/alloc-test-mpa.cc
+++ b/rt-test/alloc-test-mpa.cc
@@ -8,6 +8,7 @@
 #include "user/bmalloc.h"
 
 void main() {
+    return;
   for (int i = 4096; i < 1 << 16; ++i) {
     void *ptr = malloc(i);
     assert(ptr);
